plugins/inputs/mavlink/mavlink.go

```diff
@@ -0,0 +1,196 @@
+	if err != nil {
+		// Use default port if we could not parse out the port.
+		host = u.Host
+		port = "14550"
+	}
+
+	switch u.Scheme {
+	case "serial":
+		// Serial client
+		// Parse serial URL by hand, because it is not a compliant URL.
+		baudRate := 57600
+		device, rate, found := strings.Cut(strings.TrimPrefix(s.URL, "serial://"), ":")
+		if found {
+			r, err := strconv.Atoi(rate)
+			if err != nil {
+				return fmt.Errorf("serial baud rate not valid: %w", err)
+			}
+			baudRate = r
+		}
+
+		s.endpointConfig = []gomavlib.EndpointConf{
+			gomavlib.EndpointSerial{
+				Device: device,
+				Baud:   baudRate,
+			},
+		}
+	case "tcp":
+		if len(host) > 0 {
+			s.endpointConfig = []gomavlib.EndpointConf{
+				gomavlib.EndpointTCPClient{
+					Address: host + ":" + port,
+				},
+			}
+		} else {
+			s.endpointConfig = []gomavlib.EndpointConf{
+				gomavlib.EndpointTCPServer{
+					Address: ":" + port,
+				},
+			}
+		}
+
+	case "udp":
+		if len(host) > 0 {
+			s.endpointConfig = []gomavlib.EndpointConf{
+				gomavlib.EndpointUDPClient{
+					Address: host + ":" + port,
+				},
+			}
+		} else {
+			s.endpointConfig = []gomavlib.EndpointConf{
+				gomavlib.EndpointUDPServer{
+					Address: ":" + port,
+				},
+			}
+		}
+
+	default:
+		return fmt.Errorf("could not parse url %s", s.URL)
+	}
+
+	// Compile filter
+	s.filter, err = filter.Compile(s.FilterPattern)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Mavlink) Start(acc telegraf.Accumulator) error {
+	// Start MAVLink endpoint
+	connection, err := gomavlib.NewNode(gomavlib.NodeConf{
+		Endpoints:              s.endpointConfig,
+		Dialect:                ardupilotmega.Dialect,
+		OutVersion:             gomavlib.V2,
+		OutSystemID:            s.SystemID,
+		StreamRequestEnable:    s.StreamRequestEnable,
+		StreamRequestFrequency: s.StreamRequestFrequency,
+	})
+	if err != nil {
+		return &internal.StartupError{
+			Err:   fmt.Errorf("connecting to mavlink endpoint %s failed: %w", s.URL, err),
+			Retry: true,
+		}
+	}
+	s.connection = connection
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	s.cancel = cancelFunc
+
+	// Start routine to connect to Mavlink and stream out data async
+	go func(ctx context.Context) {
+		defer s.connection.Close()
+		if ctx.Err() != nil {
+			return
+		}
+
+		// Process MAVLink messages
+		// Use reflection to retrieve and handle all message types.
+		// (There are several hundred Mavlink message types)
+		for evt := range s.connection.Events() {
+			if ctx.Err() != nil {
+				return
+			}
+			switch evt := evt.(type) {
+			case *gomavlib.EventFrame:
+				result := convertEventFrameToMetric(evt, s.filter)
+
+				if result != nil {
+					result.AddTag("source", s.URL)
+					acc.AddMetric(result)
+				}
+
+			case *gomavlib.EventChannelOpen:
+				s.Log.Debugf("Mavlink channel opened")
+
+			case *gomavlib.EventChannelClose:
+				s.Log.Debugf("Mavlink channel closed")
+			}
+		}
+	}(ctx)
```

Comment:
This construct does not allow to stop the plugin until the next event is sent by the controller! How about
```suggestion
	go func() {
		defer s.connection.Close()
		defer s.wg.Done()

		for {
			select {
			case <-ctx.Done():
				return
			case evt := <- s.connection.Events():
				switch evt := evt.(type) {
				case *gomavlib.EventFrame:
					if result := convertEventFrameToMetric(evt, s.filter); result != nil {
						result.AddTag("source", s.URL)
						acc.AddMetric(result)
					}
				case *gomavlib.EventChannelOpen:
					s.Log.Debugf("Mavlink channel opened")
				case *gomavlib.EventChannelClose:
					s.Log.Debugf("Mavlink channel closed")
			}
		}
	}()
```

Reply:
Added the waitgroup and adopted your change